sql/src/main/java/io/crate/operation/projectors/MergeProjector.java

```diff
@@ -0,0 +1,258 @@
+/*
+ * Licensed to CRATE Technology GmbH ("Crate") under one or more contributor
+ * license agreements.  See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.  Crate licenses
+ * this file to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.operation.projectors;
+
+import com.google.common.collect.Ordering;
+import io.crate.core.collections.Buckets;
+import io.crate.core.collections.Row;
+import io.crate.core.collections.RowN;
+import io.crate.operation.RowDownstream;
+import io.crate.operation.RowDownstreamHandle;
+import io.crate.operation.RowUpstream;
+import io.crate.operation.projectors.sorting.OrderingByPosition;
+import org.elasticsearch.common.Nullable;
+
+import java.util.*;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class MergeProjector implements Projector  {
+
+    private final Ordering<Row> ordering;
+    private List<MergeProjectorDownstreamHandle> downstreamHandles = new ArrayList<>();
+    private final AtomicReference<Throwable> upstreamFailure = new AtomicReference<>(null);
+
+    private RowDownstreamHandle downstreamContext;
+
+    private Row lowestToEmit = null;
+
+    private boolean downstreamAborted = false;
```

Comment:
afaik this will be accessed from multiple threads, so at least use volatile or AtomicBoolean